### DISCUSSION TITLE
What data structure does Neovim use to handle text?

### DISCUSSION THREAD

**Discussion by ghost** [neovim/neovim#25647] [Q&A] [answered]
To store and handle the text being edited, you can use ropes, piece tables and gap buffers, or just a simple list of lines. What does Neovim use?

**Comment by justinmk** [comment]
A tree of line segments https://github.com/neovim/neovim/blob/a63c67005b9ea17214d86391e2fd649658c1bdec/src/nvim/memline.c#L8-L35

The central idea is in [ml_find_line](https://github.com/neovim/neovim/blob/a63c67005b9ea17214d86391e2fd649658c1bdec/src/nvim/memline.c#L2855).

**Reply by ghost** [reply]
Thanks. Is there any particular reason you decided to make a custom structure instead of using a rope or a gap buffer? In [this](https://github.com/neovim/neovim/wiki/Architectural-musing-and-ideas) wiki page it says Vim uses a sequence of lines, and it also talks about those different data structures.

**Reply by bfredl** [reply, marked as answer]
The main problem, which would make the data structure harder to change, is that all code around the codebase assumes that any single line would be a continuous `char *` string. This of course doesn't work well with long lines, like files with minified js or xml or JSON on a single line.

"sequence of lines" is a bit of an oversimplification. multiple small lines will be packed together into a continuous "data block"  with an additional index to quickly lookup the beginning of a line. These blocks then form a very wide tree structure (using "pointer blocks"), not exactly a B-tree rope but similar enough. This works very well with files that grow large with many small or medium sized lines (i e think any typical giant code file where lines might vary but rarely would be longer than 100-200 bytes).

I've some loose plans of doing refactors to the data structure, but this can only really be motivated with long line support (otherwise it works too "good enough" already), which can only be done with also refactoring large parts of the codebase, so it is a bit of a mess. careful planning will be required.

**Reply by ghost** [reply]
If you were to refactor it with a different data structure, would you make it a rope or a gap buffer? I think I'll mark this as the answer, but thank you justinmk as well.

**Reply by bfredl** [reply]
This is not decided, but a rope _would_ be closest to what we already have.

`memline.c` also has a small scratch buffer where multiple small edits to the same line can be performed, before it is synchronized back to the big block. not exactly gap buffer but it fills a similar role (multiple edits at the same position should be cheap).

**Reply by ghost** [reply]
Thanks for explaining! I'm going to assume since this thoughtfully planned editor uses a rope-like structure then a gap buffer isn't best. I heard Emacs has issues with jumping the cursor and it uses a gap buffer.
